File: Labs/Lab6/Lab6.java
package Labs.Lab6;

import java.util.LinkedList;
import java.util.Queue;

public class Lab6 {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] indegree = new int[numCourses];
        for (int[] pair : prerequisites) {
            indegree[pair[0]]++;
        }

        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                queue.add(i);
            }
        }

        int count = 0;
        while (!queue.isEmpty()) {
            int current = queue.poll();
            count++;
            for (int[] pair : prerequisites) {
                if (pair[1] == current) {
                    indegree[pair[0]]--;
                    if (indegree[pair[0]] == 0) {
                        queue.add(pair[0]);
                    }
                }
            }
        }

        return count == numCourses;
    }
}

File: Labs/Lab6/Lab6.py
class Lab6:
    def canFinish(self, numCourses, prerequisites):
        indegree = [0] * numCourses
        for a, b in prerequisites:
            indegree[a] += 1

        queue = [i for i in range(numCourses) if indegree[i] == 0]

        count = 0
        while queue:
            current = queue.pop(0)
            count += 1
            for a, b in prerequisites:
                if b == current:
                    indegree[a] -= 1
                    if indegree[a] == 0:
                        queue.append(a)

        return count == numCourses
